core/cont/inc/RangeCast.h

```diff
@@ -0,0 +1,185 @@
+{
+   return true;
+}
+
+template <typename>
+constexpr bool hasBeginEnd(...)
+{
+   return false;
+}
+
+} // namespace Internal
+} // namespace ROOT
+
+template <typename T, typename WrappedIterator_t, bool isDynamic>
+class TypedIter {
+
+public:
+   TypedIter(WrappedIterator_t const &iter) : _iter{iter} {}
+
+   TypedIter &operator++()
+   {
+      ++_iter;
+      return *this;
+   }
+   TypedIter operator++(int)
+   {
+      TypedIter tmp(*this);
+      operator++();
+      return tmp;
+   }
+   bool operator==(const TypedIter &rhs) const { return _iter == rhs._iter; }
+   bool operator!=(const TypedIter &rhs) const { return _iter != rhs._iter; }
+
+   void swap(TypedIter &other) { _iter.swap(other._iter); }
+
+   // We want to know at compile time whether dynamic_cast or static_cast is
+   // used. First of all to avoid overhead, but also to avoid a compiler
+   // error when using dynamic_cast on a non-polymorphic class. In C++17,
+   // this can be done easily with `if constexpr`, but for the older
+   // standards we have to use a more verbose alternative. Both ways are
+   // explicitely implemented for different standards, so that when the
+   // minimum C++ standard for ROOT is raised to C++17 it's easy to remember
+   // that we can avoid much boilerplate code in this file.
+#if (__cplusplus < 201700L)
+   T operator*() { return ROOT::Internal::Cast<T, isDynamic>::cast(*_iter); }
+#else
+   T operator*()
+   {
+      if constexpr (isDynamic) {
+         return dynamic_cast<T>(*_iter);
+      } else {
+         assert(dynamic_cast<T>(*_iter));
+         return static_cast<T>(*_iter);
+      }
+   }
+#endif
+   typename std::remove_reference<T>::type *operator->() { return &**this; }
+
+private:
+   WrappedIterator_t _iter;
+};
+
+template <typename T, typename RangeReference_t, typename ConstIter_t, bool isDynamic>
+class RangeCast {
+
+public:
+   using Range_t = typename std::remove_reference<RangeReference_t>::type;
+
+   RangeCast(RangeReference_t &&inputRange) : _inputRange{inputRange} {}
+
+   using const_iterator = TypedIter<T, ConstIter_t, isDynamic>;
+   const_iterator begin() const { return static_cast<Range_t const &>(_inputRange).begin(); }
+   const_iterator end() const { return static_cast<Range_t const &>(_inputRange).end(); }
+
+   // The non-const `begin` and `end` methods are disabled if the RangeCast wraps a const type.
+   template <typename R = Range_t, typename = std::enable_if_t<!std::is_const<R>::value>>
+   TypedIter<T, typename R::iterator, isDynamic> begin()
+   {
+      return _inputRange.begin();
+   }
+   template <typename R = Range_t, typename = std::enable_if_t<!std::is_const<R>::value>>
+   TypedIter<T, typename R::iterator, isDynamic> end()
+   {
+      return _inputRange.end();
+   }
+
+private:
+   RangeReference_t _inputRange;
+};
+
+template <typename T, typename Range_t,
+          typename ConstIter_t = typename std::remove_reference<Range_t>::type::const_iterator>
```

Comment:
Consider using the `decltype` notation I suggested above to enable this also on C-style arrays

Reply:
Okay!